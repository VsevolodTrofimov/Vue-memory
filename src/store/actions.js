import config from '@/src/utility/gameConfig'
import { preloadImages } from '@/src/utility/preload'

export const loadDeck = ({commit, dispatch}) => {
  const loading = preloadImages(config.cards.options)
  commit('deckLoaderRegister', loading)

  return new Promise((resovle, reject) => {
    loading
      .then(() => {
        commit('deckLoaded')
        resovle()
      })
      .catch(reject)
  })
}

export const startGame = ({commit, dispatch, getters}) => {
  if(getters.deck.loaded) {
    commit('fillBoard')
    commit('setScore', 0)
    commit('setMatched', 0)
    commit('showAllCards')
    commit('setStage', 'game')
  
    dispatch('commitDelayed', {
      mutation: 'maskAllCards',
      delay: config.initialVisibleTime,
      id: 'startGame-1'
    })

    //for future
    preloadImages(['/static/cards/back.png', '/static/banners/EndGame@2x.png'])
  } else {
    commit('setStage', 'loading')
    return new Promise((resolve, reject) => {
      getters.deck.loading
        .then(() => {
          dispatch('startGame')
          resolve()
        })
        .catch(reject)
    })
  }
}


export const restartGame = ({commit, dispatch}) => {
  dispatch('startGame')
}


export const endGame = ({commit}) => {
  commit('setStage', 'end')
}


export const flipCard = ({commit, dispatch, getters}, card) => {
  commit('applyHiding')
  if((card.flipped && getters.flipped.length !== 2) || card.hidden) return

  if(getters.flipped.length === 1) {
    commit('flipCard', card)
    dispatch('matchCards')
  } else {
    commit('maskAllCards')
    commit('flipCard', card)
  }
}


export const matchCards = ({commit, dispatch, getters}) => {
  if(getters.flipped[0].id === getters.flipped[1].id) {
    dispatch('applyCardMatch')
  } else {
    const diff = config.calcScore.failure(getters.matched, getters.unmatched)
    commit('setScore', getters.score + diff)
  }
}


export const applyCardMatch = ({commit, dispatch, getters}) => {
  const diff = config.calcScore.success(getters.matched, getters.unmatched)
  commit('setScore', getters.score + diff)
  
  commit('setMatched', getters.matched + 2)

  getters.flipped.forEach(card => commit('queueCardToHide', card))

  if(getters.unmatched === 0) {
    dispatch('endGame')
  } else {
    dispatch('commitDelayed', {
      mutation: 'applyHiding',
      delay: config.matchedVisibleTime,
      id: 'applyCardMatch-1'
    })
  }
}


export const commitDelayed = ({commit}, {mutation, delay, id, payload}) => {
  if(typeof id === 'undefined') {
    console.warn(
      `Not passing id parameter to commitDelayed puts you at risk
       of unknowingly resetting needed timeout`
    )
  }


  commit('delayMutation', {
    name: `${mutation}---${id}`,
    cb: () => commit(mutation, payload), // mutations cannot commit
    delay
  })
}