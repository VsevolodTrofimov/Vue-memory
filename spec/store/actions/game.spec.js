import {snapAction, MockPromise, resetTimetable} from 'vuex-snapshot'

import * as actions from '@/src/store/actions.js'


describe('Game actions', () => {
  describe('Flip card', () => {
    it('Matches "flipping first card" snapshot', () => {
      const getters = {flipped: Array(0)}
      const card = {flipped: false, url: 'AS'}

      expect(
        snapAction(actions.flipCard, {getters, payload: card})
      ).toMatchSnapshot()
    })

    it('Matches "flipping second card" snapshot', () => {
      const getters = {flipped: Array(1)}
      const card = {flipped: false, url: 'AS'}

      expect(
        snapAction(actions.flipCard, {getters, payload: card})
      ).toMatchSnapshot()
    })

    it('Matches "2 flipped cards -> again" snapshot', () => {
      const getters = {flipped: Array(2)}

      const card = {flipped: true, url: 'AS'}
      expect(
        snapAction(actions.flipCard, {getters, payload: card})
      ).toMatchSnapshot()

      card.flipped = false
      expect(
        snapAction(actions.flipCard, {getters, payload: card})
      ).toMatchSnapshot()
    })

    
    // inactive states
    it('Matches "flipping same card" snapshot', () => {
      const getters = {flipped: Array(1)}
      const card = {flipped: true, url: 'AS'}

      expect(
        snapAction(actions.flipCard, {getters, payload: card})
      ).toMatchSnapshot()
    })


    it('Matches "card is hidden" snapshot', () => {
      const getters = {flipped: Array(0)}
      const card = {flipped: true, url: 'AS', hidden: true}

      expect(
        snapAction(actions.flipCard, {getters, payload: card})
      ).toMatchSnapshot()
    })
  })

  describe('Match cards', () => {
    it('Matches "match" snapshot', () => {
      const getters = {
        flipped: [{id: 1}, {id: 1}]
      }
    
      expect(snapAction(actions.matchCards, {getters})).toMatchSnapshot()
    })

    it('Matches "mismatch" snapshot', () => {
      const getters = {
        flipped: [{id: 1}, {id: 2}]
      }
    
      expect(snapAction(actions.matchCards, {getters})).toMatchSnapshot()
    })
  })

  describe('Apply card match', () => {
    it('Matches "default" snapshot', () => {
      const getters = {
        score: 150,
        matched: 10,
        unmatched: 10,
        flipped: ['card1', 'card2']
      }
    
      expect(snapAction(actions.applyCardMatch, {getters})).toMatchSnapshot()
    })

    it('Matches "last pair" snapshot', () => {
      // should think about expecting getters to update state
      const getters = {
        score: 150,
        matched: 10,
        unmatched: 0,
        flipped: ['card1', 'card2']
      }
    
      expect(snapAction(actions.applyCardMatch, {getters})).toMatchSnapshot()
    })

  })
})