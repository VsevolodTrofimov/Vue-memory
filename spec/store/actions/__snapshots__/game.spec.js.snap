// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game actions Apply card match Matches "default" snapshot 1`] = `
"COMMIT: setScore
570
---
COMMIT: setMatched
12
---
COMMIT: queueCardToHide
\\"card1\\"
---
COMMIT: queueCardToHide
\\"card2\\"
---
DISPATCH: commitDelayed
{\\"mutation\\":\\"applyHiding\\",\\"delay\\":500,\\"id\\":\\"applyCardMatch-1\\"}
---
"
`;

exports[`Game actions Apply card match Matches "last pair" snapshot 1`] = `
"COMMIT: setScore
150
---
COMMIT: setMatched
12
---
COMMIT: queueCardToHide
\\"card1\\"
---
COMMIT: queueCardToHide
\\"card2\\"
---
DISPATCH: endGame
---
"
`;

exports[`Game actions Flip card Matches "2 flipped cards -> again" snapshot 1`] = `
"COMMIT: applyHiding
---
COMMIT: maskAllCards
---
COMMIT: flipCard
{\\"flipped\\":true,\\"url\\":\\"AS\\"}
---
"
`;

exports[`Game actions Flip card Matches "2 flipped cards -> again" snapshot 2`] = `
"COMMIT: applyHiding
---
COMMIT: maskAllCards
---
COMMIT: flipCard
{\\"flipped\\":false,\\"url\\":\\"AS\\"}
---
"
`;

exports[`Game actions Flip card Matches "card is hidden" snapshot 1`] = `
"COMMIT: applyHiding
---
"
`;

exports[`Game actions Flip card Matches "flipping first card" snapshot 1`] = `
"COMMIT: applyHiding
---
COMMIT: maskAllCards
---
COMMIT: flipCard
{\\"flipped\\":false,\\"url\\":\\"AS\\"}
---
"
`;

exports[`Game actions Flip card Matches "flipping same card" snapshot 1`] = `
"COMMIT: applyHiding
---
"
`;

exports[`Game actions Flip card Matches "flipping second card" snapshot 1`] = `
"COMMIT: applyHiding
---
COMMIT: flipCard
{\\"flipped\\":false,\\"url\\":\\"AS\\"}
---
DISPATCH: matchCards
---
"
`;

exports[`Game actions Match cards Matches "match" snapshot 1`] = `
"DISPATCH: applyCardMatch
---
"
`;

exports[`Game actions Match cards Matches "mismatch" snapshot 1`] = `
"COMMIT: setScore
null
---
"
`;
